superset/utils/lock.py

```diff
@@ -0,0 +1,88 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import json
+import logging
+import uuid
+from collections.abc import Iterator
+from contextlib import contextmanager
+from datetime import datetime, timedelta
+from typing import Any, cast, TypeVar, Union
+
+from superset.exceptions import CreateKeyValueDistributedLockFailedException
+from superset.key_value.exceptions import KeyValueCreateFailedError
+from superset.key_value.types import KeyValueResource, PickleKeyValueCodec
+
+LOCK_EXPIRATION = timedelta(seconds=30)
+logger = logging.getLogger(__name__)
+
+
+def serialize(params: dict[str, Any]) -> str:
+    """
+    Serialize parameters into a string.
+    """
+
+    T = TypeVar(
+        "T",
+        bound=Union[dict[str, Any], list[Any], int, float, str, bool, None],
+    )
+
+    def sort(obj: T) -> T:
+        if isinstance(obj, dict):
+            return cast(T, {k: sort(v) for k, v in sorted(obj.items())})
+        if isinstance(obj, list):
+            return cast(T, [sort(x) for x in obj])
+        return obj
+
+    return json.dumps(params)
+
+
+@contextmanager
+def KeyValueDistributedLock(  # pylint: disable=invalid-name
+    namespace: str,
+    **kwargs: Any,
+) -> Iterator[uuid.UUID]:
+    """
+    KV global lock for refreshing tokens.
+    """
+    # pylint: disable=import-outside-toplevel
+    from superset.commands.key_value.create import CreateKeyValueCommand
+    from superset.commands.key_value.delete import DeleteKeyValueCommand
+    from superset.commands.key_value.delete_expired import DeleteExpiredKeyValueCommand
+
+    key = uuid.uuid5(uuid.uuid5(uuid.NAMESPACE_DNS, namespace), serialize(kwargs))
+    logger.debug("Acquiring lock on namespace %s for key %s", namespace, key)
+    try:
+        DeleteExpiredKeyValueCommand(resource=KeyValueResource.LOCK).run()
+        CreateKeyValueCommand(
+            resource=KeyValueResource.LOCK,
+            codec=PickleKeyValueCodec(),
```

Comment:
the dreaded pickle, hard to imagine an attack vector on this case

Reply:
Yeah, the stored value is just `True`, so it should be fine.